validator/src/lib.rs

```diff
@@ -279,20 +279,35 @@ fn validate_temporal<E: Interpreter + Clone + Display>(
 
     // Calculate epsilon
     env.epsilon = Rational::from(i64::MAX);
-    let mut prev_timepoint = None;
-    for (timepoint, _) in span_actions_map.iter() {
-        if let Some(prev_timepoint) = prev_timepoint {
+    let mut prev_action_and_timepoint: Option<(&SpanAction<E>, &Rational)> = None;
+    let mut epsilon_explanation = None;
+
+    for (timepoint, action) in span_actions_map.iter() {
+        if let Some((prev_action, prev_timepoint)) = prev_action_and_timepoint {
             let diff = timepoint.clone() - prev_timepoint;
             if diff < env.epsilon {
                 env.epsilon = diff;
+
+                // The span actions are built in order to match an effect
+                // so the following effects should not be `None`.
+                let prev_effect = prev_action.effects().first();
+                let effect = action.effects().first();
+                debug_assert!(prev_effect.is_some());
+                debug_assert!(effect.is_some());
+                epsilon_explanation = Some((prev_effect.unwrap(), effect.unwrap()));
             }
         }
-        prev_timepoint = Some(timepoint);
+        prev_action_and_timepoint = Some((action, timepoint));
     }
     if let Some(min_epsilon) = min_epsilon {
         if env.epsilon < *min_epsilon {
+            // The explanation is set if `env.epsilon` is set.
```

Comment:
Why not put the `if env.epsilon < *min_epsilon` directly in for loop (at ` env.epsilon = diff;`). That would greatly simplify the logic.